baseclass/src/main/java/com/google/apigee/calloutbaseclass/CalloutsBase.java

```diff
@@ -0,0 +1,207 @@
+import java.io.StringWriter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * <p>https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * <p>Base Class for Apigee Java Callouts containing commonly used methods across callouts.
+ */
+public abstract class CalloutsBase {
+
+  private static final String commonErrorPatternString = "^(.+?)[:;] (.+)$";
+  private static final String variableReferencePatternString =
+      "(.*?)\\{([^\\{\\} :][^\\{\\} ]*?)\\}(.*?)";
+
+  private static final String ERROR_FLOW_VARIABLE = "error";
+  private static final String EXCEPTION_FLOW_VARIABLE = "exception";
+  private static final String EXCEPTION_STACKTRACE_FLOW_VARIABLE = "exception_stacktrace";
+  private static final String LOG_FLOW_VARIABLE = "log";
+
+  private static final Pattern commonErrorPattern = Pattern.compile(commonErrorPatternString);
+  private static final Pattern variableReferencePattern =
+      Pattern.compile(variableReferencePatternString);
+
+  private final Map<String, String> properties;
+
+  public CalloutsBase(Map properties) {
+    this.properties = convertMap(properties);
+  }
+
+  /**
+   * Convert untyped map to HashMap with String keys and values
+   *
+   * @param properties untyped map
+   * @return HashMap with String keys and values from untyped map
+   */
+  private static Map<String, String> convertMap(Map properties) {
+    Map<String, String> stringHashMap = new HashMap<>();
+    for (Object key : properties.keySet()) {
+      Object value = properties.get(key);
+      if ((key instanceof String) && (value instanceof String)) {
+        stringHashMap.put((String) key, (String) value);
+      }
+    }
+    return Collections.unmodifiableMap(stringHashMap);
+  }
+
+  /**
+   * Retrieves a required property from properties map. Returns null if not found in properties map.
+   *
+   * @param propertyName Name of property to retrieve value
+   * @param messageContext Message Context
+   * @return Value of property in properties map
+   */
+  public String getOptionalProperty(String propertyName, MessageContext messageContext) {
+    if (!this.properties.containsKey(propertyName)) {
+      return null;
+    }
+    String value = this.properties.get(propertyName);
+    value = resolveVariableReferences(value.trim(), messageContext);
+    if (value.equals("")) {
```

Comment:
"".equals(value) is in general better because we know "" cannot be null already.  NPE is a pretty serious problem in production servers.